lib/IRGen/GenThunk.cpp

```diff
@@ -52,11 +54,49 @@ IRGenModule::getAddrOfDispatchThunk(SILDeclRef declRef,
 
 static FunctionPointer lookupMethod(IRGenFunction &IGF,
                                     SILDeclRef declRef) {
-  // Find the witness table.
-  llvm::Value *wtable = (IGF.CurFn->arg_end() - 1);
+  // Get the thunk type, which is the same as the type of the protocol
+  // requirement or vtable entry, and collect parameters.
+  auto &constantInfo = IGF.IGM.getSILModule().Types.getConstantInfo(declRef);
+  Signature signature = IGF.IGM.getSignature(constantInfo.SILFnType);
 
-  // Find the witness we're interested in.
-  return emitWitnessMethodValue(IGF, wtable, declRef);
+  auto *decl = cast<AbstractFunctionDecl>(declRef.getDecl());
+  auto *dc = decl->getDeclContext();
+
+  // Protocol case.
+  if (auto *proto = dyn_cast<ProtocolDecl>(dc)) {
+    // Find the witness table.
+    llvm::Value *wtable = (IGF.CurFn->arg_end() - 1);
+
+    // Find the witness we're interested in.
+    return emitWitnessMethodValue(IGF, wtable, declRef);
+  }
+
+  // Class case.
+  auto *classDecl = cast<ClassDecl>(dc);
+
+  // Load the metadata, or use the 'self' value if we have a static method.
+  llvm::Value *self = (IGF.CurFn->arg_end() - 1);
```

Comment:
Hey @rjmccall, this PR is not ready for review yet (needs more tests, and the same refactoring I did on the witness method side to kill the duplicate code paths here), but I have a question. How do I get the 'self' parameter of a class method here? It's always the last argument in the AST type and SIL type, but we might have additional parameters in IRGen because the method itself might be generic. CurFn->arg_end() is clearly the wrong thing to use here.

Reply:
The self parameter of a class method will end up getting positioned after all the others, I think, so that it falls into the context parameter.  The only thing that goes after it is the error parameter.

Reply:
Ok, thanks. I'll add a comment with a FIXME to make it more robust in the future (but of course once we have a stable ABI we can't change parameter order anyway).